File: backend/app/router/service.py
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.deps.auth import CurrentUser
from app.deps.roles import require_roles
from app.models.service import ServiceGuest, ServiceTableWine, TableStatus
from app.schemas.service import (
    TableCreate,
    TablePatch,
    TableDetail,
    TableListResponse,
    TableListItem,
    StepAdvanceResponse,
    GuestCreate,
    GuestPatch,
    GuestOut,
    WineEntryCreate,
    WineEntryPatch,
    WineEntryOut,
)
from app.crud import service as crud

router = APIRouter(prefix="/api/service", tags=["service"])


# --- role policy (tune as you like) ---
CAN_VIEW = ("server", "expo", "sommelier", "manager")
CAN_TABLE_EDIT = ("expo", "sommelier", "manager")
CAN_STEPS = ("expo", "sommelier", "manager")
CAN_WINES = ("sommelier", "manager")
CAN_GUESTS = ("server", "expo", "sommelier", "manager")


@router.get("/tables", response_model=TableListResponse, dependencies=[Depends(require_roles(*CAN_VIEW))])
def list_tables(
    db: Session = Depends(get_db),
    status: TableStatus = Query(TableStatus.OPEN),
    page: int = Query(1, ge=1),
    limit: int = Query(25, ge=1, le=100),
    updated_since: Optional[str] = Query(None),
):
    dt = None
    if updated_since:
        try:
            dt = datetime.fromisoformat(updated_since)
        except Exception:
            raise HTTPException(status_code=400, detail="updated_since must be ISO datetime")

    total, items = crud.list_tables(db, status=status, page=page, limit=limit, updated_since=dt)
    return TableListResponse(
        items=[
            TableListItem(
                id=t.id,
                table_number=t.table_number,
                location=t.location,
                status=t.status,
                step_index=t.step_index,
                guest_count=t.guest_count,
                arrived_at=t.arrived_at,
                seated_at=t.seated_at,
                updated_at=t.updated_at,
            )
            for t in items
        ],
        page=page,
        limit=limit,
        total=total,
    )


@router.post("/tables", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_TABLE_EDIT))])
def create_table(
    payload: TableCreate,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_TABLE_EDIT)),
):
    t = crud.create_table(db, payload.table_number, payload.location, payload.guest_count, payload.notes)
    # reload full
    t = crud.get_table(db, t.id)
    return t


@router.get("/tables/{table_id}", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_VIEW))])
def get_table_detail(table_id: str, db: Session = Depends(get_db)):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    return t


@router.patch("/tables/{table_id}", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_TABLE_EDIT))])
def patch_table(
    table_id: str,
    payload: TablePatch,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_TABLE_EDIT)),
):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    data = payload.dict(exclude_unset=True)
    t = crud.patch_table(db, t, data, actor_user_id=user.id)
    return t


@router.post("/tables/{table_id}/arrive", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_TABLE_EDIT))])
def arrive(table_id: str, db: Session = Depends(get_db), user: CurrentUser = Depends(require_roles(*CAN_TABLE_EDIT))):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    return crud.mark_arrived(db, t, actor_user_id=user.id)


@router.post("/tables/{table_id}/seat", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_TABLE_EDIT))])
def seat(table_id: str, db: Session = Depends(get_db), user: CurrentUser = Depends(require_roles(*CAN_TABLE_EDIT))):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    return crud.mark_seated(db, t, actor_user_id=user.id)


@router.post("/tables/{table_id}/complete", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_TABLE_EDIT))])
def complete(table_id: str, db: Session = Depends(get_db), user: CurrentUser = Depends(require_roles(*CAN_TABLE_EDIT))):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    return crud.complete_table(db, t, actor_user_id=user.id)


# ---------- steps ----------
@router.post("/tables/{table_id}/next", response_model=StepAdvanceResponse, dependencies=[Depends(require_roles(*CAN_STEPS))])
def next_step(table_id: str, db: Session = Depends(get_db), user: CurrentUser = Depends(require_roles(*CAN_STEPS))):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    t = crud.next_step(db, t, actor_user_id=user.id)
    return StepAdvanceResponse(table_id=t.id, step_index=t.step_index, updated_at=t.updated_at)


@router.post("/tables/{table_id}/undo", response_model=StepAdvanceResponse, dependencies=[Depends(require_roles(*CAN_STEPS))])
def undo_step(table_id: str, db: Session = Depends(get_db), user: CurrentUser = Depends(require_roles(*CAN_STEPS))):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    t = crud.undo_step(db, t, actor_user_id=user.id)
    return StepAdvanceResponse(table_id=t.id, step_index=t.step_index, updated_at=t.updated_at)


# ---------- guests ----------
@router.post("/tables/{table_id}/guests", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_GUESTS))])
def add_guest(
    table_id: str,
    payload: GuestCreate,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_GUESTS)),
):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")
    _, t = crud.add_guest(db, t, payload.dict(exclude_unset=True), actor_user_id=user.id)
    return t


@router.patch("/tables/{table_id}/guests/{guest_id}", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_GUESTS))])
def patch_guest(
    table_id: str,
    guest_id: str,
    payload: GuestPatch,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_GUESTS)),
):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")

    g = db.query(ServiceGuest).filter(ServiceGuest.id == guest_id, ServiceGuest.table_id == table_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="Guest not found")

    _, t = crud.update_guest(db, t, g, payload.dict(exclude_unset=True), actor_user_id=user.id)
    return t


@router.delete("/tables/{table_id}/guests/{guest_id}", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_GUESTS))])
def delete_guest(
    table_id: str,
    guest_id: str,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_GUESTS)),
):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")

    g = db.query(ServiceGuest).filter(ServiceGuest.id == guest_id, ServiceGuest.table_id == table_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="Guest not found")

    t = crud.remove_guest(db, t, g, actor_user_id=user.id)
    return t


# ---------- wines (locked until arrived) ----------
@router.post("/tables/{table_id}/wines", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_WINES))])
def add_wine(
    table_id: str,
    payload: WineEntryCreate,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_WINES)),
):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")

    if not crud.ensure_wines_unlocked(t):
        raise HTTPException(status_code=409, detail="Wines are locked until arrival")

    _, t = crud.add_wine(db, t, payload.dict(exclude_unset=True), actor_user_id=user.id)
    return t


@router.patch("/tables/{table_id}/wines/{wine_entry_id}", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_WINES))])
def patch_wine(
    table_id: str,
    wine_entry_id: str,
    payload: WineEntryPatch,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_WINES)),
):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")

    if not crud.ensure_wines_unlocked(t):
        raise HTTPException(status_code=409, detail="Wines are locked until arrival")

    w = db.query(ServiceTableWine).filter(ServiceTableWine.id == wine_entry_id, ServiceTableWine.table_id == table_id).first()
    if not w:
        raise HTTPException(status_code=404, detail="Wine entry not found")

    _, t = crud.update_wine(db, t, w, payload.dict(exclude_unset=True), actor_user_id=user.id)
    return t


@router.delete("/tables/{table_id}/wines/{wine_entry_id}", response_model=TableDetail, dependencies=[Depends(require_roles(*CAN_WINES))])
def delete_wine(
    table_id: str,
    wine_entry_id: str,
    db: Session = Depends(get_db),
    user: CurrentUser = Depends(require_roles(*CAN_WINES)),
):
    t = crud.get_table(db, table_id)
    if not t:
        raise HTTPException(status_code=404, detail="Table not found")

    if not crud.ensure_wines_unlocked(t):
        raise HTTPException(status_code=409, detail="Wines are locked until arrival")

    w = db.query(ServiceTableWine).filter(ServiceTableWine.id == wine_entry_id, ServiceTableWine.table_id == table_id).first()
    if not w:
        raise HTTPException(status_code=404, detail="Wine entry not found")

    t = crud.remove_wine(db, t, w, actor_user_id=user.id)
    return t

File: frontend/src/api/service.ts
import { api } from "./client";

export type TableStatus = "OPEN" | "COMPLETED" | "CANCELED";
export type WineKind = "BOTTLE" | "BTG";

export type TableListItem = {
  id: string;
  table_number: string;
  location?: string | null;
  status: TableStatus;
  step_index: number;
  guest_count: number;
  arrived_at?: string | null;
  seated_at?: string | null;
  updated_at: string;
};

export type Guest = {
  id: string;
  name?: string | null;
  allergy?: string | null;
  protein_sub?: string | null;
  doneness?: string | null;
  substitutions?: string | null;
  notes?: string | null;
  updated_at: string;
};

export type WineEntry = {
  id: string;
  kind: WineKind;
  wine_id?: string | null;
  label: string;
  quantity: number;
  updated_at: string;
};

export type StepEvent = {
  id: string;
  event_type: string;
  from_step?: number | null;
  to_step?: number | null;
  payload?: string | null;
  actor_user_id?: string | null;
  created_at: string;
};

export type TableDetail = {
  id: string;
  table_number: string;
  location?: string | null;
  status: TableStatus;
  step_index: number;
  guest_count: number;
  notes?: string | null;
  arrived_at?: string | null;
  seated_at?: string | null;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
  guests: Guest[];
  wines: WineEntry[];
  events: StepEvent[];
};

export type TableListResponse = {
  items: TableListItem[];
  page: number;
  limit: number;
  total: number;
};

export const serviceApi = {
  listTables: (params: { status?: TableStatus; page?: number; limit?: number; updated_since?: string } = {}) => {
    const q = new URLSearchParams();
    if (params.status) q.set("status", params.status);
    if (params.page) q.set("page", String(params.page));
    if (params.limit) q.set("limit", String(params.limit));
    if (params.updated_since) q.set("updated_since", params.updated_since);
    return api.get<TableListResponse>(`/api/service/tables?${q.toString()}`);
  },

  getTable: (tableId: string) => api.get<TableDetail>(`/api/service/tables/${tableId}`),

  createTable: (payload: { table_number: string; location?: string; guest_count?: number; notes?: string }) =>
    api.post<TableDetail>(`/api/service/tables`, payload),

  patchTable: (tableId: string, payload: Partial<{ table_number: string; location: string; guest_count: number; notes: string }>) =>
    api.patch<TableDetail>(`/api/service/tables/${tableId}`, payload),

  arrive: (tableId: string) => api.post<TableDetail>(`/api/service/tables/${tableId}/arrive`),
  seat: (tableId: string) => api.post<TableDetail>(`/api/service/tables/${tableId}/seat`),
  complete: (tableId: string) => api.post<TableDetail>(`/api/service/tables/${tableId}/complete`),

  next: (tableId: string) => api.post<{ table_id: string; step_index: number; updated_at: string }>(`/api/service/tables/${tableId}/next`),
  undo: (tableId: string) => api.post<{ table_id: string; step_index: number; updated_at: string }>(`/api/service/tables/${tableId}/undo`),

  addGuest: (tableId: string, payload: Partial<Guest>) => api.post<TableDetail>(`/api/service/tables/${tableId}/guests`, payload),
  patchGuest: (tableId: string, guestId: string, payload: Partial<Guest>) =>
    api.patch<TableDetail>(`/api/service/tables/${tableId}/guests/${guestId}`, payload),
  deleteGuest: (tableId: string, guestId: string) => api.del<TableDetail>(`/api/service/tables/${tableId}/guests/${guestId}`),

  addWine: (tableId: string, payload: { kind: WineKind; wine_id?: string; label: string; quantity: number }) =>
    api.post<TableDetail>(`/api/service/tables/${tableId}/wines`, payload),
  patchWine: (tableId: string, wineEntryId: string, payload: Partial<{ label: string; quantity: number }>) =>
    api.patch<TableDetail>(`/api/service/tables/${tableId}/wines/${wineEntryId}`, payload),
  deleteWine: (tableId: string, wineEntryId: string) => api.del<TableDetail>(`/api/service/tables/${tableId}/wines/${wineEntryId}`),
};
